rest/backend/src/main/scala/io/udash/rest/server/ExposesREST.scala

```diff
@@ -0,0 +1,126 @@
+  def urlPartToRaw(raw: String, isStringArg: Boolean): RawValue
+
+  def handleRestCall(getterChain: List[RawInvocation], invocation: RawInvocation)(implicit ec: ExecutionContext): Future[String] = {
+    try {
+      val receiver = rawLocalRpc.resolveGetterChain(getterChain)
+      receiver.call(invocation.rpcName, invocation.argLists).map(rawToString)
+    } catch {
+      case ex: Exception =>
+        Future.failed(ex)
+    }
+  }
+
+  def parseHttpRequest(req: HttpServletRequest, httpMethod: Class[_ <: RESTMethod]): (List[RawInvocation], RawInvocation) = {
+    val invocations = List.newBuilder[RawInvocation]
+    val path: Array[String] = Option(req.getPathInfo).map(_.stripPrefix("/").split("/")).getOrElse(Array.empty[String])
+    val bodyContent = req.getReader.lines().toArray.mkString("\n")
+    lazy val bodyValues = read[Map[String, framework.RawValue]](stringToRaw(bodyContent))(bodyValuesReader)
+
+    def findRestParamName(annotations: Seq[MetadataAnnotation]): Option[String] =
+      annotations.find(_.isInstanceOf[RESTParamName]).map(_.asInstanceOf[RESTParamName].restName)
+
+    def parseInvocations(path: Seq[String], metadata: RPCMetadata[_]): Unit = {
+      if (path.isEmpty) throw ExposesREST.NotFound(req.getPathInfo)
+
+      val methodName = path.head
+      var nextParts = path.tail
+
+      if (!metadata.signatures.contains(methodName))
+        throw ExposesREST.NotFound(req.getPathInfo)
+
+      val methodMetadata = metadata.signatures(methodName)
+      var hasBodyArgs = false
+
+      val args: List[List[RawValue]] = methodMetadata.paramMetadata.map { argsList =>
+        argsList.map { arg =>
+          val argTypeAnnotations = arg.annotations.filter(_.isInstanceOf[ArgumentType])
+          argTypeAnnotations.headOption match {
+            case Some(_: Header) =>
+              val argName = findRestParamName(arg.annotations).getOrElse(arg.name)
+              val headerValue = req.getHeader(argName)
+              if (headerValue == null) throw ExposesREST.MissingHeader(argName)
+              headerArgumentToRaw(headerValue, arg.typeMetadata == framework.SimplifiedType.StringType)
+            case Some(_: URLPart) =>
+              if (nextParts.isEmpty) throw ExposesREST.MissingURLPart(arg.name)
+              val v = nextParts.head
+              nextParts = nextParts.tail
+              urlPartToRaw(v, arg.typeMetadata == framework.SimplifiedType.StringType)
+            case Some(_: BodyValue) =>
+              val argName = findRestParamName(arg.annotations).getOrElse(arg.name)
+              if (!bodyValues.contains(argName)) throw ExposesREST.MissingBodyValue(arg.name)
+              hasBodyArgs = true
+              bodyValues(argName)
+            case Some(_: Body) =>
+              if (bodyContent.isEmpty) throw ExposesREST.MissingBody(arg.name)
+              hasBodyArgs = true
+              stringToRaw(bodyContent)
+            case _ => // Query is a default argument type
```

Comment:
You could use `Some(_: Query) | None` to be a little more explicit. Especially needed if you opt for exhaustive match checking mentioned above.